File: sort/bucket_sort/r/bucket_sort.r
creatingBuckets <- function(arr) {
  o <- list(c(), c(), c(), c(), c(), c(), c(), c(), c(), c())
  for (i in 1:length(arr)) {
    n <- as.integer(floor(10 * arr[i]))
    o[[n]] <- c(o[[n]], arr[i])
  }
  for (i in 1:length(o)) {
    if(length(o[[i]])>1){
      o[[i]] <- insertionSort(o[[i]])
    }
  }
  mergeList(o)
}
insertionSort <- function(a) {
  for (c in 2:(length(a))) {
    d = c
    while (d > 1 && (a[d] < a[d-1])) {
      t = a[d]
      a[d] = a[d - 1]
      a[d - 1] = t
      d = d - 1
    }
  }
  a
}
mergeList <- function(q) {
  res <- as.vector(c())
  for (k in 1:length(q)) {
    if (!is.null(q[[k]]))
      res <- c(res, q[[k]])
  }
  print("Vector after BucketSort")
  print(res)
}
arr <- c(0.78, 0.17, 0.39, 0.26)
creatingBuckets(arr)

File: sort/bucket_sort/ruby/bucket_sort.rb
def bucket_sort(array)
  if array.empty?
    return
  end
  
  bucket_size = array.length

  # Determine minimum and maximum values
  min_value = array.min
  max_value = array.max

  # Initialise buckets
  bucket_count = ((max_value - min_value) / bucket_size).floor + 1
  buckets = Array.new(bucket_count)
  (0...buckets.length).each do |i|
    buckets[i] = []
  end

  # Distribute input array values into buckets
  (0...array.length).each do |i|
    buckets[((array[i] - min_value) / bucket_size).floor].push(array[i])
  end

  # Sort buckets and place back into input array
  array.clear
  (0...buckets.length).each do |i|
    buckets[i].sort!
    buckets[i].each do |value|
      array.push(value)
    end
  end
end

array = [8, 5, 3, 1, 9, 6, 0, 7, 4, 2, 5]
puts "Before: ", array.inspect
bucket_sort(array)
puts "After:", array.inspect
